python/taichi/lang/field.py

```diff
@@ -0,0 +1,459 @@
+        """
+        raise NotImplementedError()
+
+    @python_scope
+    def __getitem__(self, key):
+        """Gets field element in Python scope.
+
+        Args:
+            key (Union[List[int], int, None]): Coordinates of the field element.
+
+        Returns:
+            element type: Value retrieved.
+        """
+        raise NotImplementedError()
+
+    def __str__(self):
+        if impl.inside_kernel():
+            return self.__repr__()  # make pybind11 happy, see Matrix.__str__
+        else:
+            return str(self.to_numpy())
+
+    def pad_key(self, key):
+        if key is None:
+            key = ()
+        if not isinstance(key, (tuple, list)):
+            key = (key, )
+        assert len(key) == len(self.shape)
+        return key + ((0, ) * (_ti_core.get_max_num_indices() - len(key)))
+
+    def initialize_host_accessors(self):
+        if self.host_accessors:
+            return
+        impl.get_runtime().materialize()
+        self.host_accessors = [
+            SNodeHostAccessor(e.ptr.snode()) for e in self.vars
+        ]
+
+
+class ScalarField(Field):
+    """Taichi scalar field with SNode implementation.
+
+    Args:
+        var (Expr): Field member.
+    """
+    def __init__(self, var):
+        super().__init__([var])
+
+    @python_scope
+    def fill(self, val):
+        from taichi.lang.meta import fill_tensor
+        fill_tensor(self, val)
+
+    @python_scope
+    def to_numpy(self, dtype=None):
+        if dtype is None:
+            dtype = to_numpy_type(self.dtype)
+        import numpy as np
+        arr = np.zeros(shape=self.shape, dtype=dtype)
+        from taichi.lang.meta import tensor_to_ext_arr
+        tensor_to_ext_arr(self, arr)
+        ti.sync()
+        return arr
+
+    @python_scope
+    def to_torch(self, device=None):
+        import torch
+        arr = torch.zeros(size=self.shape,
+                          dtype=to_pytorch_type(self.dtype),
+                          device=device)
+        from taichi.lang.meta import tensor_to_ext_arr
+        tensor_to_ext_arr(self, arr)
+        ti.sync()
+        return arr
+
+    @python_scope
+    def from_numpy(self, arr):
+        assert len(self.shape) == len(arr.shape)
+        for i in range(len(self.shape)):
+            assert self.shape[i] == arr.shape[i]
+        if hasattr(arr, 'contiguous'):
+            arr = arr.contiguous()
+        from taichi.lang.meta import ext_arr_to_tensor
+        ext_arr_to_tensor(arr, self)
+        ti.sync()
+
+    @python_scope
+    def __setitem__(self, key, value):
+        self.initialize_host_accessors()
+        self.host_accessors[0].setter(value, *self.pad_key(key))
+
+    @python_scope
+    def __getitem__(self, key):
+        self.initialize_host_accessors()
+        return self.host_accessors[0].getter(*self.pad_key(key))
+
+    def __repr__(self):
+        # make interactive shell happy, prevent materialization
+        return '<ti.field>'
+
+
+class MatrixField(Field):
```

Comment:
nit: I feel like it's better to put this inside `matrix.py`. WDYT?

Reply:
Maybe? Then people can add more types and add a corresponding field class in the same file?

Reply:
I think it's natural to put different field impls in separate files.

Reply:
Yes it makes sense. Then is it clearer to have a `field.py`, a `scalar_field.py`, and `matrix_field.py`?

Reply:
I'd also inclined to putting each type of field in a separate file.

Reply:
`MatrixField` is now put in `matrix.py` to avoid potential circular import.